### PR TITLE
fts-squat: fix Corrupted squat uidlist bug

### PR DESCRIPTION
**Pull request by nikwrt**
Hello,

I have experienced the "Corrupted squat uidlist" bug described in these two threads:

http://dovecot.org/list/dovecot/2012-April/135162.html
http://dovecot.org/list/dovecot/2016-January/102864.html

and I have tested a patch based on the explanation in the second thread. This has been working flawlessly for approximately a month. I'd really like to see this merged so I don't have to repatch it every time.

### CODE REVIEW

**Comment by nikwrt** [comment]
Hi, I have been running this now for approximately six months, with no issues whatsoever. Please, consider merging it.

**Comment by cmouse** [comment]
I'll try to have a look at this next week. Sorry for the delay.

**Comment by cmouse** [comment]
We'll do internal review for this

File: src/plugins/fts-squat/fts-backend-squat.c [hunk 1]
```diff
@@ -137,6 +147,9 @@ fts_backend_squat_set_box(struct squat_fts_backend *backend,
 	if (backend->full_len != 0)
 		squat_trie_set_full_len(backend->trie, backend->full_len);
 	backend->box = box;
+	ret = squat_trie_open(backend->trie);
+	if (ret < 0)
+		return -1;
```
Comment (sirainen) [inline src/plugins/fts-squat/fts-backend-squat.c]: This is missing return value for ret >= 0 case.

Comment (nikwrt) [inline src/plugins/fts-squat/fts-backend-squat.c]: This is intended. If ret >= 0 the next line will return what the original code did:

return squat_trie_get_last_uid(backend->trie, last_uid_r);

Previously, squat_trie_open could fail and squat_trie_get_last_uid would still be called, originating the error.

Hope this helps...

Comment (sirainen) [inline src/plugins/fts-squat/fts-backend-squat.c]: But you are missing return value from the function entirely when ret >= 0, so it'll now more or less randomly return something. I'd guess the above should be: return squat_trie_open(backend->trie);

Comment (nikwrt) [inline src/plugins/fts-squat/fts-backend-squat.c]: I am puzzled, are we looking at the same code? Please click maybe on "View full changes". If ret >=0, the next line will execute, and that is:

`return squat_trie_get_last_uid(backend->trie, last_uid_r);`

Comment (sirainen) [inline src/plugins/fts-squat/fts-backend-squat.c]: You're looking at fts_backend_squat_get_last_uid(), I'm looking at fts_backend_squat_set_box().

Comment (nikwrt) [inline src/plugins/fts-squat/fts-backend-squat.c]: Ah yes, I see it now. You are right. It should return 0 and I overlooked it.

Comment (nikwrt) [inline src/plugins/fts-squat/fts-backend-squat.c]: Just wondering, should I fix it in my fork and resubmit it? Just let me know.

Comment (sirainen) [inline src/plugins/fts-squat/fts-backend-squat.c]: Yeah, that'd be easier for us :)

Comment (nikwrt) [inline src/plugins/fts-squat/fts-backend-squat.c]: It's done now and I see it's been picked up automatically by the pull request. Thanks for pointing it out.

**Comment by sirainen** [comment]
merged

**Pull request closed without merging**